user-contrib/Ltac2/Lazy.v

```
@@ -0,0 +1,36 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *         Copyright INRIA, CNRS and contributors             *)
+(* <O___,, * (see version control and CREDITS file for authors & dates) *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+Require Import Ltac2.Init.
+
+(** Type of a reference cell, similar to OCaml's ['a Lazy.t] type. *)
+Ltac2 Type 'a lazy := {
+  mutable value : 'a option;
+  thunk : unit -> 'a;
```

Why a pair instead of something like
~~~coq
Ltac2 Type 'a lazy_data := [ Evaluated ('a) | Thunk (unit -> 'a) ].
Ltac2 Type 'a lazy := 'a lazy_data ref.
~~~
?

Notably using a pair means the thunk can't be freed by the gc even after it's been evaluated.

Also ocaml when evaluating a lazy first sets to thunk to `fun () -> raise Lazy.Undefined` which provides loop detection, and if it produces an exception it sets the thunk to `fun () -> raise the_exn`. It may be worth doing this in ltac2 too (although there's the question of if we want to `zero` or `throw` for Undefined).

(ocaml also has some GC trickery so that `of_val foo` and `foo` have the same representation a lot of the time, but I don't think it's worth doing that for ltac2)

---

I think that this should be a primitive type inherited from OCaml.

---

Why?

---

@SkySkimmer are you suggesting something like the following for the implementation of `force`?
```coq
Ltac2 force : 'a t -> 'a := fun r =>
  match Ref.get r with
  | Value v => v
  | Thunk f =>
      Ref.set r (Thunk (fun _ => Control.throw Undefined));
      let run () :=
        let v := f () in
        Ref.set r (Value v); v
      in
      let handle e :=
        Ref.set r (Thunk (fun _ => Control.zero e));
        Control.zero e
      in
      Control.plus run handle
  end.
```
If yes, I would definitely use `throw` for `Undefined`.

---

It might need to be Control.case instead of plus but otherwise yes.

---

I'll close this thread since I fixed the implementation, but feel free to open another thread to explain why you think a primitive type inherited from OCaml would be better @ppedrot.